File: ios/Classes/FlutterEThree.swift
//
//  FlutterEThree.swift
//  e3kit
//
//  Created by Matheus Cardoso on 11/12/19.
//

import VirgilE3Kit
import VirgilSDK
import VirgilCrypto

struct FlutterEThree {
    let instance: EThree
    let channel: FlutterMethodChannel

    func invoke(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        switch call.method {
        case "getIdentity": getIdentity(result)
        case "hasLocalPrivateKey": hasLocalPrivateKey(result)
        case "register": register(result)
        case "rotatePrivateKey": rotatePrivateKey(result)
        case "cleanUp": cleanUp(result)
        case "findUsers": findUsers(
            try call.getArgument("identities"),
            result
        )
        case "encrypt": encrypt(
            text: try call.getArgument("text"),
            for: call.getOptionalArgument("users"),
            result
        )
        case "decrypt": decrypt(
            text: try call.getArgument("text"),
            from: call.getOptionalArgument("user"),
            result
        )
        case "backupPrivateKey": backupPrivateKey(
            password: try call.getArgument("password"),
            result
        )
        case "resetPrivateKeyBackup": resetPrivateKeyBackup(result)
        case "changePassword": changePassword(
            from: try call.getArgument("oldPassword"),
            to: try call.getArgument("newPassword"),
            result
        )
        case "restorePrivateKey": restorePrivateKey(
            password: try call.getArgument("password"),
            result
        )
        case "unregister": unregister(result)
        case "createGroup": createGroup(
            id: try call.getArgument("groupId"),
            with: try call.getArgument("users"),
            result
        )
        case "loadGroup": loadGroup(
            id: try call.getArgument("groupId"),
            initiator: try call.getArgument("initiator"),
            result
        )
        case "getGroup": getGroup(
            id: try call.getArgument("groupId"),
            result
        )
        case "deleteGroup": deleteGroup(
            id: try call.getArgument("groupId"),
            result
        )
        default:
            result(FlutterError(
                code: "method_not_recognized",
                message: "Method is not recognized",
                details: "Method name: '\(call.method)'"
            ))
        }
    }

    func getIdentity(_ result: @escaping FlutterResult) {
        result(instance.identity)
    }

    func hasLocalPrivateKey(_ result: @escaping FlutterResult) {
        do {
            result(try instance.hasLocalPrivateKey())
        } catch let error {
            return result(error.toFlutterError())
        }
    }

    func register(_ result: @escaping FlutterResult) {
        instance.register(completion: { error in
            if let error = error {
                return result(error.toFlutterError())
            }

            return result(true)
        })
    }

    func rotatePrivateKey(_ result: @escaping FlutterResult) {
        instance.rotatePrivateKey(completion: { error in
            if let error = error {
                return result(error.toFlutterError())
            }

            return result(true)
        })
    }

    func cleanUp(_ result: @escaping FlutterResult) {
        do {
            try instance.cleanUp()
            return result(true)
        } catch let error {
            return result(error.toFlutterError())
        }
    }

    func findUsers(_ identities: [String], _ result: @escaping FlutterResult) {
        instance.findUsers(with: identities).start(completion: { res, error in
            if let error = error {
                return result(FlutterError(
                    code: "find_users_failed",
                    message: "Failed to find users",
                    details: error.localizedDescription
                ))
            }

            guard let res = res else {
                return result(FlutterError(
                    code: "find_users_failed",
                    message: "Failed to find users",
                    details: "Result is null"
                ))
            }

            do {
                let res = try res.compactMapValues({
                    try $0.getRawCard().exportAsBase64EncodedString()
                })

                return result(res)
            } catch let error {
                return result(FlutterError(
                    code: "find_users_failed",
                    message: "Could not encode result",
                    details: error.localizedDescription
                ))
            }
        })
    }

    func encrypt(
        text: String,
        for users: [String: String]? = nil,
        _ result: @escaping FlutterResult
    ) {
        do {
            let users = try users?.mapValues {
                try instance.cardManager.importCard(fromBase64Encoded: $0)
            }

            result(try instance.encrypt(text: text, for: users))
        } catch let error {
            return result(error.toFlutterError())
        }
    }

    func decrypt(
        text: String,
        from user: String?,
        _ result: @escaping FlutterResult
    ) {
        do {
            if let user = user {
                let card = try instance
                    .cardManager
                    .importCard(fromBase64Encoded: user)

                return result(try instance.decrypt(text: text, from: card))
            }

            return result(try instance.decrypt(text: text))
        } catch let error {
            return result(error.toFlutterError())
        }
    }

    func backupPrivateKey(
        password: String,
        _ result: @escaping FlutterResult
    ) {
        instance.backupPrivateKey(password: password).start { res in
            switch res {
            case .success:
                return result(true)
            case .failure(let error):
                return result(error.toFlutterError())
            }
        }
    }

    func resetPrivateKeyBackup(_ result: @escaping FlutterResult) {
        instance.resetPrivateKeyBackup().start { res in
            switch res {
            case .success:
                return result(true)
            case .failure(let error):
                return result(error.toFlutterError())
            }
        }
    }

    func changePassword(
        from oldPassword: String,
        to newPassword: String,
        _ result: @escaping FlutterResult
    ) {
        instance.changePassword(from: oldPassword, to: newPassword).start { res in
            switch res {
            case .success:
                return result(true)
            case .failure(let error):
                return result(error.toFlutterError())
            }
        }
    }

    func restorePrivateKey(
        password: String,
        _ result: @escaping FlutterResult
    ) {
        instance.restorePrivateKey(password: password).start { res in
            switch res {
            case .success:
                return result(true)
            case .failure(let error):
                return result(error.toFlutterError())
            }
        }
    }

    func unregister(_ result: @escaping FlutterResult) {
        instance.unregister().start { res in
            switch res {
            case .success:
                return result(true)
            case .failure(let error):
                return result(error.toFlutterError())
            }
        }
    }

    func _registerGroup(
        _ group: Group,
        _ result: @escaping FlutterResult
    ) {
        let uuid = "GROUP:\(UUID().uuidString)"
        groups[uuid] = FlutterGroup(
            origin: instance,
            instance: group,
            channel: self.channel
        )
        return result(uuid)
    }

    func _handleGroupResult(
        _ res: Result<Group, Error>,
        _ result: @escaping FlutterResult
    ) {
        switch res {
        case .success(let group):
            _registerGroup(group, result)
        case .failure(let error):
            return result(error.toFlutterError())
        }
    }

    func createGroup(
        id: String,
        with users: [String: String],
        _ result: @escaping FlutterResult
    ) {
        do {
            let users = try users.mapValues {
                try instance
                    .cardManager
                    .importCard(fromBase64Encoded: $0)
            }

            instance.createGroup(id: id, with: users).start { res in
                self._handleGroupResult(res, result)
            }
        } catch let error {
            return result(error.toFlutterError())
        }
    }

    func loadGroup(
        id: String,
        initiator: String,
        _ result: @escaping FlutterResult
    ) {
        do {
            let card = try instance
                .cardManager
                .importCard(fromBase64Encoded: initiator)

            instance.loadGroup(id: id, initiator: card).start { res in
                self._handleGroupResult(res, result)
            }
        } catch let error {
            return result(error.toFlutterError())
        }
    }

    func getGroup(
        id: String,
        _ result: @escaping FlutterResult
    ) {
        do {
            if let group = try instance.getGroup(id: id) {
                _registerGroup(group, result)
            } else {
                result(FlutterError(
                    code: "group_null",
                    message: "Group is null",
                    details: nil
                ))
            }
        } catch let error {
            return result(error.toFlutterError())
        }
    }

    func deleteGroup(
        id: String,
        _ result: @escaping FlutterResult
    ) {
        instance.deleteGroup(id: id).start { res in
            switch(res) {
            case .success:
                result(true)
            case .failure(let error):
                result(error.toFlutterError())
            }
        }
    }
}

File: android/src/main/kotlin/com/virgilsecurity/e3kit/FlutterEThree.kt
package com.virgilsecurity.e3kit

import android.app.Activity
import android.os.AsyncTask
import android.os.Looper
import com.virgilsecurity.android.common.model.FindUsersResult
import com.virgilsecurity.android.ethree.interaction.EThree
import com.virgilsecurity.common.callback.OnCompleteListener
import com.virgilsecurity.common.callback.OnResultListener
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel


class FlutterEThree {
    val instance: EThree
    val channel: MethodChannel
    val activity: Activity

    constructor(instance: EThree, channel: MethodChannel, activity: Activity) {
        this.instance = instance
        this.channel = channel
        this.activity = activity
    }

    fun invoke(call: MethodCall, result: MethodChannel.Result) {
        fun <T: Any>getArgument(argument: String, optional: Boolean = false): T {
            val arg = (call.arguments as? HashMap<String, Any>)?.getValue(argument) as? T

            if(!optional && arg == null) {
                val errorCode = "argument_not_found"
                val errorMessage = "Could not find argument `$argument` of type ${arg?.javaClass?.name}"
                result.error(errorCode, errorMessage, null)
                throw Error("$errorCode: $errorMessage")
            }

            return arg!!
        }

        fun <T: Any>getOptionalArgument(argument: String): T? {
            return try {
                getArgument(argument, true)
            } catch(e: Throwable) {
                null
            }
        }


        try {
            when (call.method) {
                "getIdentity" -> getIdentity(result)
                "hasLocalPrivateKey" -> hasLocalPrivateKey(result)
                "register" -> register(result)
                "rotatePrivateKey" -> rotatePrivateKey(result)
                "cleanUp" -> cleanUp(result)
                "findUsers" -> findUsers(
                        getArgument("identities"),
                        result
                )
                "encrypt" -> encrypt(
                        getArgument("text"),
                        getOptionalArgument("users"),
                        result
                )
                "decrypt" -> decrypt(
                        getArgument("text"),
                        getOptionalArgument("user"),
                        result
                )
                "backupPrivateKey" -> backupPrivateKey(
                    getArgument("password"),
                    result
                )
                "resetPrivateKeyBackup" -> resetPrivateKeyBackup(result)
                "changePassword" -> changePassword(
                    getArgument("oldPassword"),
                    getArgument("newPassword"),
                    result
                )
                "restorePrivateKey" -> restorePrivateKey(
                        getArgument("password"),
                        result
                )
                "unregister" -> unregister(result)
                else -> activity.runOnUiThread {
                    result.error(
                            "method_not_recognized",
                            "Method is not recognized",
                            "Method name: '${call.method}'"
                    )
                }
            }
        } catch(e: Throwable) {
            result.error(e.toFlutterError())
        }
    }

    private fun getIdentity(result: MethodChannel.Result) {
        activity.runOnUiThread {
            result.success(instance.identity)
        }
    }

    private fun hasLocalPrivateKey(result: MethodChannel.Result) {
        activity.runOnUiThread {
            result.success(instance.hasLocalPrivateKey())
        }
    }

    private fun register(result: MethodChannel.Result) {
        instance.register().addCallback(object: OnCompleteListener {
            override fun onSuccess() {
                activity.runOnUiThread {
                    result.success(true)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }

    private fun rotatePrivateKey(result: MethodChannel.Result) {
        instance.rotatePrivateKey().addCallback(object: OnCompleteListener {
            override fun onSuccess() {
                activity.runOnUiThread {
                    result.success(true)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }

    private fun cleanUp(result: MethodChannel.Result) {
        AsyncTask.execute {
            try {
                instance.cleanup()
            } catch(e: Throwable) {
                activity.runOnUiThread {
                    result.error(e.toFlutterError())
                }
                
                return@execute
            }

            activity.runOnUiThread {
                result.success(true)
            }
        }
    }

    private fun findUsers(identities: List<String>, result: MethodChannel.Result) {
        instance.findUsers(identities, true).addCallback(object: OnResultListener<FindUsersResult> {
            override fun onSuccess(res: FindUsersResult) {
                val mapped = res.mapValues {
                    it.value.rawCard.exportAsBase64String()!!
                }

                activity.runOnUiThread {
                    result.success(mapped)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }

    private fun encrypt(
            text: String,
            users: HashMap<String, String>? = null,
            result: MethodChannel.Result
    ) {
        val mapped = users?.mapValues {
            instance.cardManager.importCardAsString(it.value)!!
        }

        val findUsersResult = if (mapped != null) FindUsersResult(mapped) else null

        val res = instance.encrypt(text, findUsersResult)

        result.success(res)
    }

    private fun decrypt(
            text: String,
            user: String? = null,
            result: MethodChannel.Result
    ) {
        val imported = instance.cardManager.importCardAsString(user)
        val res = instance.decrypt(text, imported)

        result.success(res)
    }

    private fun backupPrivateKey(password: String, result: MethodChannel.Result) {
        instance.backupPrivateKey(password).addCallback(object: OnCompleteListener {
            override fun onSuccess() {
                activity.runOnUiThread {
                    result.success(true)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }

    private fun resetPrivateKeyBackup(result: MethodChannel.Result) {
        instance.resetPrivateKeyBackup().addCallback(object: OnCompleteListener {
            override fun onSuccess() {
                activity.runOnUiThread {
                    result.success(true)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }

    private fun changePassword(oldPassword: String, newPassword: String, result: MethodChannel.Result) {
        instance.changePassword(oldPassword, newPassword).addCallback(object: OnCompleteListener {
            override fun onSuccess() {
                activity.runOnUiThread {
                    result.success(true)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }

    private fun restorePrivateKey(password: String, result: MethodChannel.Result) {
        instance.restorePrivateKey(password).addCallback(object: OnCompleteListener {
            override fun onSuccess() {
                activity.runOnUiThread {
                    result.success(true)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }

    private fun unregister(result: MethodChannel.Result) {
        instance.unregister().addCallback(object: OnCompleteListener {
            override fun onSuccess() {
                activity.runOnUiThread {
                    result.success(true)
                }
            }
            override fun onError(throwable: Throwable) {
                activity.runOnUiThread {
                    result.error(throwable.toFlutterError())
                }
            }
        })
    }
}